midend/predication.cpp

```diff
@@ -13,11 +13,120 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
-
 #include "predication.h"
 #include "frontends/p4/cloner.h"
 
+
 namespace P4 {
+/// convert an expression into a string that uniqely identifies the value referenced
+/// return null cstring if not a reference to a constant thing.
+static cstring expr_name(const IR::Expression *exp) {
+    if (auto p = exp->to<IR::PathExpression>())
+        return p->path->name;
+    if (auto m = exp->to<IR::Member>()) {
+        if (auto base = expr_name(m->expr))
+            return base + "." + m->member;
+    } else if (auto a = exp->to<IR::ArrayIndex>()) {
+        if (auto k = a->right->to<IR::Constant>())
+            if (auto base = expr_name(a->left))
+                return base + "." + std::to_string(k->asInt()); }
+    return cstring();
+}
+
+bool Predication::EmptyStatementRemover::preorder(IR::BlockStatement * block) {
+    for (auto iter = block->components.begin(); iter != block->components.end();) {
+        if ((*iter)->is<IR::EmptyStatement>()) {
+            iter = block->components.erase(iter);
+        } else if ((*iter)->is<IR::BlockStatement>()) {
+            visit(*iter);
+            if ((*iter)->to<IR::BlockStatement>()->components.size() == 0) {
+                iter = block->components.erase(iter);
+            } else {
+                iter++;
+            }
+        } else {
+            iter++;
+        }
+    }
+    return false;
+}
+
+const IR::AssignmentStatement *
+Predication::ExpressionReplacer::preorder(IR::AssignmentStatement * statement) {
+    // fill the conditions from context
+    while (context != nullptr) {
+        if (context->node->is<IR::IfStatement>()) {
+            conditions.push_back(context->node->to<IR::IfStatement>()->condition);
+        }
+        context = context->parent;
+    }
+
+    if (!statement->right->is<IR::Mux>()) {
+        statement->right = new IR::Mux(conditions.back(), statement->left, statement->left);
+    }
+    visit(statement->right);
+    return statement;
+}
+
+void Predication::ExpressionReplacer::emplaceExpression(IR::Mux * mux) {
+    auto condition = conditions[conditions.size() - currentNestingLevel];
+    bool thenElsePass = travesalPath[currentNestingLevel - 1];
+    mux->e0 = condition;
+    if (thenElsePass) {
+        mux->e1 = rightExpression;
+    } else {
+        mux->e2 = rightExpression;
+    }
+}
+
+void Predication::ExpressionReplacer::visitThen(IR::Mux * mux) {
+    auto condition = conditions[conditions.size() - currentNestingLevel];
+    auto statement = findContext<IR::AssignmentStatement>();
+    auto leftName = expr_name(statement->left);
+    auto thenExprName = expr_name(mux->e1);
+    if (expr_name(mux->e2) == expr_name(statement->left)) {
```

Comment:
This will be equal when the result is nullptr and the two expressions are not syntactically identical.
Is this the desired behavior?